// @lwrjs/lwc-ssr/build/cjs/viewTransformer/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/lwc-ssr/src/viewTransformer/index.ts
__markAsModule(exports);
__export(exports, {
  createFetchEndowment: () => import_utils2.createFetchEndowment,
  default: () => lwcSsrViewTransformer
});
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_instrumentation = __toModule(require("@lwrjs/instrumentation"));
var import_shared_utils = __toModule(require("@lwrjs/shared-utils"));
var import_identity = __toModule(require("../identity.cjs"));
var import_ssr_bootstrap = __toModule(require("../utils/ssr-bootstrap.cjs"));
var import_utils = __toModule(require("../utils/utils.cjs"));
var import_utils2 = __toModule(require("../utils/utils.cjs"));
function lwcSsrViewTransformer(options, {config, moduleBundler, resourceRegistry}) {
  const routes = [...config.routes, ...config.errorRoutes];
  return {
    name: "ssr-lwc-transformer",
    async link(stringBuilder, viewContext, metadata) {
      if (!viewContext.view.bootstrap?.ssr) {
        return {};
      }
      import_diagnostics.logger.debug({label: "lwcSsrViewTransformer", message: "link"});
      import_diagnostics.logger.verbose({
        label: "lwcSsrViewTransformer",
        message: "link input",
        additionalInfo: stringBuilder
      });
      if (!metadata.serverData) {
        metadata.serverData = {};
      }
      if (!metadata.serverDebug) {
        metadata.serverDebug = {};
      }
      const {customElements, serverData, serverDebug} = metadata;
      const {debug} = viewContext.runtimeEnvironment;
      const ssrModules = getComponentsToSSR(customElements, stringBuilder);
      const rootSpecifiers = Object.values(ssrModules).map((m) => m.specifier);
      if (!rootSpecifiers.length) {
        return {};
      }
      const islands = rootSpecifiers.join(",");
      const specifier = `${import_identity.LWC_SSR_PREFIX}${viewContext.view.id}/${islands}`;
      const allProps = {};
      Object.entries(ssrModules).forEach(([s, info]) => allProps[s] = info.props);
      let pageTtl;
      await (0, import_instrumentation.getTracer)().trace({name: import_instrumentation.ViewSpan.RenderIsland, attributes: {specifiers: islands}}, async () => {
        const {results, errors} = await (0, import_ssr_bootstrap.runServerBootstrap)({specifier, props: allProps, serverData}, moduleBundler, resourceRegistry, routes, viewContext);
        for (const root in results) {
          const {html, props, cache} = results[root];
          const {tagName, startOffset, endOffset, hydrate} = ssrModules[root];
          pageTtl = (0, import_shared_utils.shortestTtl)(cache?.ttl, pageTtl);
          if (html) {
            let propsAttr = "";
            if (hydrate) {
              const propsId = (0, import_identity.getPropsId)();
              propsAttr = ` ${import_identity.SSR_PROPS_ATTR}="${propsId}"`;
              serverData[propsId] = props;
            }
            const [, remain] = html.split(`<${tagName}`);
            stringBuilder.overwrite(startOffset, endOffset, [`<${tagName}`, propsAttr, remain].join(""));
          }
        }
        results && addPreloadLinks(results, stringBuilder);
        errors && handleErrors(errors, customElements, islands, debug, serverDebug);
      });
      import_diagnostics.logger.verbose({
        label: "lwcSsrViewTransformer",
        message: "response",
        additionalInfo: stringBuilder
      });
      return {cache: {ttl: pageTtl}};
    }
  };
}
function getComponentsToSSR(customElements, stringBuilder) {
  const cmpInfo = {};
  for (const {tagName, location, props: rawProps = {}} of customElements.values()) {
    const isCsr = (0, import_shared_utils.isCsrIsland)(rawProps);
    if (isCsr && location) {
      const {startOffset, endOffset} = location;
      stringBuilder.overwrite(startOffset, endOffset, `<${tagName}></${tagName}>`);
    }
    if (!isCsr && !rawProps?.["lwc:external"] && location) {
      const {startOffset, endOffset} = location;
      const specifier = (0, import_shared_utils.kebabCaseToModuleSpecifier)(tagName);
      const hydrate = (0, import_shared_utils.isHydrateOnLoad)(rawProps);
      const props = {...rawProps};
      delete props[import_shared_utils.HYDRATE_DIRECTIVE];
      cmpInfo[specifier] = {startOffset, endOffset, props, tagName, specifier, hydrate};
    }
  }
  return cmpInfo;
}
function addPreloadLinks(results, stringBuilder) {
  let htmlLinks = "";
  const allLinks = [];
  Object.values(results).forEach(({markup: {links = []} = {}}) => allLinks.push(...links));
  allLinks.forEach(({href, rel, as, fetchpriority}) => {
    const relStr = rel ? ` rel="${rel}"` : "", asStr = as ? ` as="${as}"` : "", fetchStr = fetchpriority ? ` fetchpriority="${fetchpriority}"` : "";
    htmlLinks += `<link href="${href}"${relStr}${asStr}${fetchStr}>
`;
  });
  if (htmlLinks) {
    const headIndex = stringBuilder.original.indexOf("</head>");
    if (headIndex >= 0) {
      stringBuilder.prependLeft(headIndex, htmlLinks);
    } else {
      import_diagnostics.logger.error("Adding links during server-side rendering failed. Could not find the </head> tag.");
    }
  }
}
function handleErrors(errors, customElements, specifiers, debug, serverDebug) {
  const allErrors = Object.values(errors).join(", ");
  if (!debug && !(0, import_shared_utils.getFeatureFlags)().SSR_WITH_CSR_FALLBACK) {
    throw new Error(allErrors);
  }
  Object.entries(errors).forEach(([specifier, err]) => {
    const ce = customElements.find(({tagName}) => specifier === (0, import_shared_utils.kebabCaseToModuleSpecifier)(tagName));
    if (ce) {
      ce.props === void 0 ? ce.props = {
        [import_shared_utils.HYDRATE_DIRECTIVE]: import_shared_utils.HYDRATE_CLIENT_VALUE
      } : ce.props[import_shared_utils.HYDRATE_DIRECTIVE] = import_shared_utils.HYDRATE_CLIENT_VALUE;
    }
    const errMessage = (0, import_utils.createSsrErrorMessage)(specifier, err);
    import_diagnostics.logger.warn(errMessage, err.stack);
  });
  if (debug) {
    serverDebug.message = (0, import_utils.createSsrErrorMessage)(specifiers, allErrors);
  }
}
