// @lwrjs/instrumentation/build/es/spans.js
export var ConfigSpan;
(function (ConfigSpan) {
    ConfigSpan["LoadConfig"] = "lwr.config.load";
    ConfigSpan["ExecuteConfigHooks"] = "lwr.config.hooks.init";
    ConfigSpan["ExecuteStartHooks"] = "lwr.config.hooks.start";
    ConfigSpan["ExecuteInstrHooks"] = "lwr.config.hooks.instrumentation";
    ConfigSpan["ResolveConfig"] = "lwr.config.resolve";
    ConfigSpan["ValidateConfig"] = "lwr.config.validate";
})(ConfigSpan || (ConfigSpan = {}));
export var CoreSpan;
(function (CoreSpan) {
    CoreSpan["CreateServer"] = "lwr.core.create";
    CoreSpan["CreateServerContext"] = "lwr.core.context";
    CoreSpan["InitServer"] = "lwr.core.init";
    CoreSpan["StartServer"] = "lwr.core.start";
    CoreSpan["UnhandledRejection"] = "lwr.core.unhandledRejection";
})(CoreSpan || (CoreSpan = {}));
export var LambdaSpan;
(function (LambdaSpan) {
    LambdaSpan["CreateHandler"] = "lwr.lambda.create";
    LambdaSpan["CreateServerContext"] = "lwr.lambda.context";
    LambdaSpan["UnhandledRejection"] = "lwr.lambda.unhandledRejection";
})(LambdaSpan || (LambdaSpan = {}));
export var RequestHandlerSpan;
(function (RequestHandlerSpan) {
    RequestHandlerSpan["GetView"] = "lwr.handle.view";
    RequestHandlerSpan["GetAsset"] = "lwr.handle.asset";
    RequestHandlerSpan["GetModule"] = "lwr.handle.module";
    RequestHandlerSpan["GetBundle"] = "lwr.handle.bundle";
    RequestHandlerSpan["GetResource"] = "lwr.handle.resource";
    RequestHandlerSpan["GetMapping"] = "lwr.handle.mapping";
})(RequestHandlerSpan || (RequestHandlerSpan = {}));
export var ViewSpan;
(function (ViewSpan) {
    ViewSpan["ExecuteRouteHandler"] = "lwr.view.routeHandler";
    ViewSpan["RenderView"] = "lwr.view.render";
    ViewSpan["ServerSideRender"] = "lwr.view.ssr";
    ViewSpan["RenderPage"] = "lwr.view.ssr.page";
    ViewSpan["RenderIsland"] = "lwr.view.ssr.island";
    ViewSpan["PreloadData"] = "lwr.view.preload.data";
    ViewSpan["GetServerData"] = "lwr.view.ssr.data";
    ViewSpan["Fetch"] = "lwr.view.ssr.fetch";
    ViewSpan["BootstrapService"] = "lwr.view.ssr.service";
    ViewSpan["RenderComponent"] = "lwr.view.ssr.renderComponent";
    ViewSpan["Generate"] = "lwr.view.generate";
    ViewSpan["ServerBundle"] = "lwr.view.ssr.bundle";
})(ViewSpan || (ViewSpan = {}));
export var BundleSpan;
(function (BundleSpan) {
    BundleSpan["GetBundle"] = "lwr.bundle.get";
    BundleSpan["Rollup"] = "lwr.bundle.rollup";
})(BundleSpan || (BundleSpan = {}));
export const DefaultSpans = new Set([
    LambdaSpan.CreateHandler,
    LambdaSpan.UnhandledRejection,
    RequestHandlerSpan.GetView,
    RequestHandlerSpan.GetAsset,
    RequestHandlerSpan.GetModule,
    RequestHandlerSpan.GetBundle,
    RequestHandlerSpan.GetResource,
    RequestHandlerSpan.GetMapping,
    ViewSpan.ExecuteRouteHandler,
    ViewSpan.RenderView,
    ViewSpan.ServerSideRender,
    ViewSpan.GetServerData,
    ViewSpan.Fetch,
    ViewSpan.RenderComponent,
    ViewSpan.BootstrapService,
]);
