// @lwrjs/lwc-ssr/build/cjs/identity.cjs
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};

// packages/@lwrjs/lwc-ssr/src/identity.ts
__markAsModule(exports);
__export(exports, {
  LWC_SSR_PREFIX: () => LWC_SSR_PREFIX,
  SSR_PROPS_ATTR: () => SSR_PROPS_ATTR,
  getPropsId: () => getPropsId,
  getSsrServices: () => getSsrServices,
  parseBootstrapSpecifier: () => parseBootstrapSpecifier
});
var LWC_SSR_PREFIX = "@lwrjs/lwc-ssr/";
var SSR_PROPS_ATTR = "data-lwr-props-id";
function getPropsId() {
  return `lwcprops${Math.floor(Math.random() * 65536).toString(16)}`;
}
function parseBootstrapSpecifier(specifier, routes = []) {
  if (specifier.startsWith(LWC_SSR_PREFIX)) {
    const parts = specifier.replace(LWC_SSR_PREFIX, "");
    const slash = parts.indexOf("/");
    if (slash) {
      const routeId = parts.substring(0, slash);
      const rootSpecifierStr = parts.substring(slash + 1);
      if (rootSpecifierStr) {
        return {
          rootSpecifiers: rootSpecifierStr.split(","),
          route: routes.find((r) => r.id === routeId)
        };
      }
    }
  }
  return void 0;
}
function getSsrServices(route) {
  return route && route.bootstrap.ssr ? route.bootstrap.services.filter((s) => s.ssr).map(({name}) => name) : [];
}
