// @lwrjs/lwc-ssr/build/cjs/dataViewTransformer/index.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/lwc-ssr/src/dataViewTransformer/index.ts
__markAsModule(exports);
__export(exports, {
  createFetchEndowment: () => import_utils.createFetchEndowment,
  default: () => preloadDataViewTransformer
});
var import_diagnostics = __toModule(require("@lwrjs/diagnostics"));
var import_instrumentation = __toModule(require("@lwrjs/instrumentation"));
var import_identity = __toModule(require("../identity.cjs"));
var import_ssr_bootstrap = __toModule(require("../utils/ssr-bootstrap.cjs"));
var import_utils = __toModule(require("../utils/utils.cjs"));
var NAME = "preload-data-transformer";
function preloadDataViewTransformer(_options, {config, moduleBundler, resourceRegistry}) {
  const routes = [...config.routes, ...config.errorRoutes];
  return {
    name: NAME,
    async link(_stringBuilder, viewContext, metadata) {
      const {rootComponent} = viewContext.view;
      if (!rootComponent || !viewContext.view.bootstrap?.preloadData || viewContext.view.bootstrap?.ssr) {
        return {};
      }
      import_diagnostics.logger.debug({label: NAME, message: `Preload data for root component "${rootComponent}"`});
      const {results = {}, errors} = await (0, import_instrumentation.getTracer)().trace({
        name: import_instrumentation.ViewSpan.PreloadData,
        attributes: {rootComponent}
      }, async () => {
        return await (0, import_ssr_bootstrap.runServerBootstrap)({specifier: `${import_identity.LWC_SSR_PREFIX}${viewContext.view.id}/${rootComponent}`, props: {}}, moduleBundler, resourceRegistry, routes, viewContext);
      });
      if (errors) {
        for (const err of Object.entries(errors)) {
          import_diagnostics.logger.warn({
            label: "preloadDataViewTransformer",
            message: `Unexpected error during preload data: ${err[0]}`
          }, err[1]);
        }
        return {};
      }
      const {props, cache: {ttl} = {ttl: void 0}} = results[rootComponent];
      import_diagnostics.logger.verbose({label: NAME, message: "response", additionalInfo: props});
      metadata.serverData = metadata.serverData || {};
      Object.assign(metadata.serverData, props);
      return {cache: {ttl}};
    }
  };
}
